src/main/java/com/exadel/discount/service/impl/FavoriteServiceImpl.java

```diff
@@ -0,0 +1,119 @@
+import org.springframework.data.domain.Pageable;
+import org.springframework.stereotype.Service;
+import org.springframework.transaction.annotation.Transactional;
+
+import java.util.List;
+import java.util.Optional;
+import java.util.UUID;
+
+@Service
+@Slf4j
+@AllArgsConstructor
+public class FavoriteServiceImpl implements FavoriteService {
+    private final FavoriteRepository favoriteRepository;
+    private final UserRepository userRepository;
+    private final FavoriteMapper favoriteMapper;
+    private final DiscountRepository discountRepository;
+
+    @Override
+    public List<FavoriteDTO> findAllFavorites(int pageNumber, int pageSize, String sortDirection, String sortField,
+                                              FavoriteFilter favoriteFilter) {
+        Pageable paging = SortPageMaker.makePageable(pageNumber, pageSize, sortDirection, sortField);
+        log.debug("Getting sorted page-list of all Favorites");
+        Page<Favorite> favoriteList = favoriteRepository
+                .findAll(preparePredicateForFindingAllFavorites(favoriteFilter), paging);
+        if (favoriteList.isEmpty()) throw
+                new NotFoundException("No favorites are found");
+
+        log.debug("Successfully got sorted page-list of all Favorites");
+        return favoriteMapper.toFavoriteDTOList(favoriteList.toList());
+    }
+
+    @Override
+    public FavoriteDTO findFavoriteById(UUID id) {
+        log.debug("Finding Favorite by ID");
+        Optional<Favorite> favoriteOptional = favoriteRepository.findById(id);
+        if (favoriteOptional.isEmpty()) throw
+                new NotFoundException(String.format("Favorite with id %s not found", id));
+        else log.debug("Successfully Favorite is found by ID");
+        return favoriteMapper.toFavoriteDTO(favoriteOptional.get());
+    }
+
+    @Transactional
+    @Override
+    public FavoriteDTO assignFavoriteToUser(CreateFavoriteDTO createFavoriteDTO) {
+        log.debug("Finding of certain user by ID");
+        User user = userRepository
+                .findById(createFavoriteDTO.getUserId())
+                .orElseThrow(() -> new NotFoundException(String
+                        .format("User with id %s not found", createFavoriteDTO.getUserId())));
+
+        Discount discount = discountRepository
+                .findById(createFavoriteDTO.getDiscountId())
+                .orElseThrow(() -> new NotFoundException(String
+                        .format("Discount with id %s not found", createFavoriteDTO.getDiscountId())));
+        log.debug("Successfully certain User and Discount are found by ID. Starting Favorite creation/saving.");
+
+        Favorite favorite = new Favorite();
+        favorite.setUser(user);
+        favorite.setDiscount(discount);
+
+        favoriteRepository.save(favorite);
+        log.debug("Successfully new Favorite is saved to certain user");
+
+        return favoriteMapper.toFavoriteDTO(favorite);
```

Comment:
return favorite that save() method returns

Reply:
Done